modules/siddhi-core/src/main/java/org/wso2/siddhi/core/aggregation/AggregationRuntime.java

```diff
@@ -431,14 +431,18 @@ public void startPurging() {
         incrementalDataPurging.executeIncrementalDataPurging();
     }
 
-    public void recreateInMemoryData() {
+    public void recreateInMemoryData(boolean isFirstEventArrived) {
         // State only updated when first event arrives to IncrementalAggregationProcessor
-        this.isFirstEventArrived = true;
-        for (Map.Entry<TimePeriod.Duration, IncrementalExecutor> durationIncrementalExecutorEntry :
-                this.incrementalExecutorMap.entrySet()) {
-            durationIncrementalExecutorEntry.getValue().setProcessingExecutor(true);
+        if (isFirstEventArrived) {
+            this.isFirstEventArrived = true;
+            for (Map.Entry<TimePeriod.Duration, IncrementalExecutor> durationIncrementalExecutorEntry :
+                    this.incrementalExecutorMap.entrySet()) {
+                durationIncrementalExecutorEntry.getValue().setProcessingExecutor(true);
+            }
+        }
+        synchronized (this) {
+            this.recreateInMemoryData.recreateInMemoryData();
```

Comment:
Why are we adding synchronized (this) instead of synchronizing the method?

Reply:
fixed in 31abdcb